common/beerocks/bcl/source/beerocks_socket_event_loop.cpp

```diff
@@ -0,0 +1,336 @@
+    LOG_IF(m_epoll_fd == -1, FATAL) << "Failed creating epoll: " << strerror(errno);
+}
+
+SocketEventLoop::~SocketEventLoop()
+{
+    // Delete all the sockets in the poll
+    LOG(DEBUG) << "Removing " << m_fd_to_event_data.size()
+               << " FDs from the event loop: " << m_fd_to_event_data;
+
+    while (!m_fd_to_event_data.empty()) {
+        auto event_data = m_fd_to_event_data.begin()->second;
+        del_socket(event_data->socket);
+    }
+
+    // Close the poll fd
+    LOG_IF(close(m_epoll_fd) == -1, ERROR)
+        << "Failed closing epoll file descriptor: " << strerror(errno);
+}
+
+bool SocketEventLoop::add_socket(std::shared_ptr<Socket> socket,
+                                 const EventLoop::EventHandlers &handlers,
+                                 std::chrono::milliseconds timeout)
+{
+    if (!socket) {
+        LOG(ERROR) << "Invalid socket pointer!";
+        return false;
+    }
+
+    // Make sure that the FD is not already part of the poll
+    if (m_fd_to_event_data.find(socket->getSocketFd()) != m_fd_to_event_data.end()) {
+        LOG(WARNING) << "Requested to add FD (" << socket->getSocketFd()
+                     << ") to the poll, but it's already there...";
+
+        return false;
+    }
+
+    // Build EventData structure
+    auto event_data      = std::make_shared<EventData>();
+    event_data->socket   = socket;
+    event_data->handlers = handlers;
+
+    // If timeout is set and a handler is defined, initialize a timerfd
+    if (timeout > std::chrono::milliseconds::zero()) {
+        if (handlers.on_timeout) {
+
+            // TODO: timerfd may not be supported on all kernels.
+            //       Implement a check at creation, and in case it's not supported
+            //       base the timeouts mechanism on the main timeout of the loop.
+
+            // Create a new timerfd file descriptor and initialize the timeout values
+            event_data->timeout_value = timeout;
+            event_data->timeout = std::make_shared<Socket>(timerfd_create(CLOCK_MONOTONIC, 0));
+
+            if (!event_data->timeout || event_data->timeout->getSocketFd() == -1) {
+                LOG(ERROR) << "Failed creating timerfd: " << strerror(errno);
+                return false;
+            }
+
+            // Convert the timeout into seconds and nano-seconds
+            auto timeout_sec = std::chrono::duration_cast<std::chrono::seconds>(timeout);
+            auto timeout_nanosec =
+                std::chrono::duration_cast<std::chrono::nanoseconds>(timeout - timeout_sec);
+
+            timespec timeout_spec({.tv_sec  = static_cast<time_t>(timeout_sec.count()),
+                                   .tv_nsec = static_cast<long>(timeout_nanosec.count())});
+            itimerspec timer_val({.it_interval = timeout_spec, .it_value = timeout_spec});
+
+            // Set the timerfd timeout value
+            if (timerfd_settime(event_data->timeout->getSocketFd(), 0, &timer_val, nullptr) == -1) {
```

Comment:
Timer should be configured with initial expiration value only, timer interval is not required and should not be set.
`timerfd_settime()` should be called here and again whenever a read or write event is triggered for the associated socket. 
If timeout is considered an error, then all handlers for the file descriptor should be removed calling `del_socket()` as with error and disconnected events. Otherwise, `timerfd_settime()` should be called to rearm timer before start waiting for events again.

Reply:
Who's responsible for re-arming the timeout in this case? EventLoop or Handler?

Reply:
As it is implemented now, the EventLoop shall rearm the timer.

In a future refactoring, we might consider moving timeout handling out of the EventLoop completely. This would give us more flexibility as it would be the clients of the EventLoop API who decide.

Reply:
Since the mechanism is already implemented I prefer to keep it, and refactor later if it won't suite our needs.